flexmeasures/data/services/utils.py

```diff
@@ -0,0 +1,90 @@
+import functools
+from flask import current_app
+from rq.job import Job
+
+import hashlib
+import base64
+
+from flexmeasures.data.models.time_series import Sensor
+
+
+def make_hash_sha256(o):
+    """
+    SHA256 instead of Python's hash function because apparently, python native hashing function
+    yields different results on restarts.
+    Source: https://stackoverflow.com/a/42151923
+    """
+    hasher = hashlib.sha256()
+    hasher.update(repr(make_hashable(o)).encode())
+    return base64.b64encode(hasher.digest()).decode()
+
+
+def make_hashable(o):
+    """
+    Function to create hashes for dictionaries with nested objects
+    Source: https://stackoverflow.com/a/42151923
+    """
+    if isinstance(o, (tuple, list)):
+        return tuple((make_hashable(e) for e in o))
+
+    if isinstance(o, dict):
+        return tuple(sorted((k, make_hashable(v)) for k, v in o.items()))
+
+    if isinstance(o, (set, frozenset)):
+        return tuple(sorted(make_hashable(e) for e in o))
+
+    if isinstance(o, Sensor):
+        return tuple(
+            (
+                make_hashable(getattr(o, attr, None))
+                for attr in ["attributes", "annotations", "id"]
```

Comment:
I chose them because the `id` identifies uniquely a sensor and `annotations` and `attributes` can change. For instance, `custom-scheduler`.

Please, let me know any other field that can trigger a new computation of the schedule.

## Can change
attributes
annotations

## Cannot change:
id (primary key)
generic_asset_id
name
unit
timezone 
event_resolution
knowledge_horizon_fnc
knowledge_horizon_par

Reply:
Thanks, I understand now. We should make it clear in the code why some properties are included and why some are not. Something like: ```
"""
In principle all properties (except ID) of a given sensor could be changed, but not all changes are relevant to warrant reanalysis (e.g. scheduling or forecasting).
"""
```
I would also move this definition to become a class method of the `Sensor` class.

I believe the annotations property refers to another database table containing annotations as a time series of string values. So I wouldn't include this one.

The attributes of the sensor's generic asset may also influence the scheduling results, because defaults for, for example, which price sensor is relevant, can be stored as an attribute of either the sensor or the generic asset. So I would add that one.